openstack/utils/choose_version.go

```diff
@@ -20,9 +21,12 @@ var goodStatus = map[string]bool{
 	"stable":    true,
 }
 
-// ChooseVersion queries the base endpoint of an API to choose the most recent non-experimental alternative from a service's
-// published versions.
-// It returns the highest-Priority Version among the alternatives that are provided, as well as its corresponding endpoint.
+// ChooseVersion queries the base endpoint of an API to choose the identity service version.
```

Comment:
You're sure this is limited to "identity service", i.e. Keystone?

Reply:
It has a reference to `IdentityBase` down the code, dunno why.

Reply:
Yes, I got fooled by the doc string into believing this could be used for Nova. I was very confused about why it never could find a good version to use until @mdbooth pointed out that it is taking the ProviderClient as input, not the ServiceClient. It is indeed only checking the identity endpoint and picks a version for Keystone.

Reply:
Given that gophercloud is working on a major release, maybe it's time to change that?

Reply:
Yeah that would be nice!

Reply:
It would also be really good to have this in the v1 branch though, if possible. We have a bunch of potential near term use cases for it, one of which is described in https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1715#issuecomment-1752923531.

Reply:
@mdbooth I thought you did not want this kind of negotiation of the version? I have reworked the CAPO PR for that reason to set an explicit version in all cases instead of using a function like this. In which cases do you think we should use this?

Reply:
I probably got confused and thought I was commenting on the PR in general :)

You are correct, I don't want to do negotiation in CAPO: we either get a specific version or we don't execute the code. There may be cases where some code might run with one of a small number of specific microversions (e.g. GET server) but we should never, e.g. request microversion N and get microversion N+1.

That said, it might be useful to be able to specify a microversion limit for testing so we can test old microversions against a newer server. The microversion requested would still be at the point of use, though, and we would still either get it or not get it.

However, CAPO hopefully isn't the only consumer of this code. Negotiation might be useful elsewhere, e.g. a CLI utility.